modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonClientExperimentalCodegen.java

```diff
@@ -970,13 +972,17 @@ public String toInstantiationType(Schema property) {
     @Override
     protected void addAdditionPropertiesToCodeGenModel(CodegenModel codegenModel, Schema schema) {
         Schema addProps = ModelUtils.getAdditionalProperties(schema);
-        if (addProps != null && addProps.get$ref() == null) {
-            // if AdditionalProperties exists and is an inline definition, get its datatype and store it in m.parent
-            String typeString = getTypeDeclaration(addProps);
-            codegenModel.additionalPropertiesType = typeString;
-        } else {
-            addParentContainer(codegenModel, codegenModel.name, schema);
+        if (addProps != null) {
+            if (addProps.get$ref() == null) {
+                // if AdditionalProperties exists and is an inline definition, get its datatype and
+                // store it in codegenModel.additionalPropertiesType.
+                codegenModel.additionalPropertiesType = getTypeDeclaration(addProps);;
+            } else {
+                addParentContainer(codegenModel, codegenModel.name, schema);
```

Comment:
@spacether , can you comment on the purpose of calling addParentContainer() for python? I don't understand the use case.
Also, previously if addProps was null, we were calling addParentContainer(). Why? addProps is null when the OAS schema has `additionalProperties: false`.

Reply:
So I'm not sure what benefit we get from calling addParentContainer. It looks like I copied and pasted that code from DefaultCodegen at:
https://github.com/OpenAPITools/openapi-generator/blob/dc1bdac82073093913b3066fca51b730f2f8cedb/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java#L2796

1.
The way I read that code is that if additional properties exist, we call the if block.
If they don't call the else block. And the else block code is consistent with what the DefaultCodegen does in its own addAdditionPropertiesToCodeGenModel method.

2. Per my #1 interpretation addParentContainer would be called if additionalProperties is false. It would also be called if additionalProperties is null.

3. I don't think that it should matter if additionalProperties is inline or a $ref. Doesn't the code that we have work for both cases?

Reply:
If additionalProperties is null and the spec version is 3.0.0 or 3.0.03 or higher, shouldn't additionalProperties default to true before we hande it here?
v3.0.3 explicitly says that it defaults to true here: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md
v3.0.0 points us to the https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md JSON schema definition, and does not tell us a default value.

Reply:
I think the default is always specified regardless of the version, because of the following text in the OAS specs:

> The following properties are taken from the JSON Schema definition but their definitions were adjusted to the OpenAPI Specification.
> additionalProperties - Value can be boolean or object. Inline or referenced schema MUST be of a Schema Object and not a standard JSON Schema.

Note OAS 2.0 is not exactly the same language as OAS 3.0, but they both mean the `additionalProperties` keyword is specified in the JSON schema validation draft.

Hence whatever is specified in JSON schema validation is applicable to OAS. Each OAS version is referring to a different version of the JSON drafts, but they all seem to have the same default behavior.

Specifically from https://tools.ietf.org/html/draft-wright-json-schema-validation-00#page-10:

> If "additionalProperties" is absent, it may be considered present with an empty schema as a value.

Which means the values can be anything.

Reply:
BTW, this sentence in 3.0.3 is superfluous: "Consistent with JSON Schema, additionalProperties defaults to true."  It's already defined in the JSON core/validation specs, hence it should not be repeated in OAS.

In fact in OAS 3.1, it was completely removed because the goal of the technical committee is for OAS to be a superset of JSON schema, i.e. supporting all JSON schema constructs as is, without the need to rephrase the JSON specification inside the OAS spec.

Reply:
> 
> The way I read that code is that if additional properties exist, we call the if block.
> If they don't call the else block. And the else block code is consistent with what the DefaultCodegen does in its own addAdditionPropertiesToCodeGenModel method.
> 

ok, let me try to refactor that code a bit so that it works the same whether it's a reference or not. I will add a unit test too to exercise the code.

Reply:
Unfortunately, I found out the OpenAPI parser for a V2 document sets `Schema.additionalProperties` to the null value even when the OAS document has `additionalProperties: true|false`
So I need to add some logic that if the doc is version 2, then assume the old behavior, because we don't know if additionalProperties should be true or false. This seems to be a bug in the parser.

Reply:
What if you added the CLI flag supportsAdditionalProperties
And if the generator has it turned on, postprocess the spec to correctly set additionalProperties to {} if the input is null
Or should that not be done, or be done in another PR?

Reply:
I am adding a legacyAdditionalPropertiesBehavior CLI flag.